enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+- Should we teach the etcd operator to remove the protection mechanism when the `ControlPlaneMachineSet` is configured
+  with this strategy?
+- To minimise risk, should we allow this strategy only on certain platforms? (Eg disallow the strategy on public clouds)
+- Do we want to name the update strategy in a way that highlights the risks associated, eg `RecreateUnsupported`?
+
+##### The OnDelete update strategy
+
+The OnDelete strategy mirrors the OnDelete strategy familiar to users from StatefulSets.
+When a change is required, any logic for rolling out the Machine will be paused until the `ControlPlaneMachineSet`
+controller observes a deleted Machine.
+When a Machine is marked for deletion, it will create a new Machine based on the current spec.
+It will then proceed with the replacement as normal.
+
+This strategy will allow for explicit control for end users to decide when their Control Plane Machines are updated.
+In particular it will also allow them to have different specs across their control plane.
+While in the normal case this is discouraged, this could be used for short periods to test new configuration before
+rolling the updated configuration to all Control Plane Machines.
+
+Note that the OnDelete strategy is similar to the RollingUpdate strategy in that it needs to be able to add new
+Machines to the cluster to function. It can be thought of as a slower, more controlled RollingUpdate strategy where the
+user decides when each Control Plane Machine is replaced, and marks it for replacement by deleting it.
+Otherwise the replacement is identical to the RollingUpdate strategy.
+
+#### Protecting etcd quorum during update operations
+
+With the introduction of the [etcd protection enhancement](https://github.com/openshift/enhancements/pull/943), the
+ControlPlaneMachineSet does not need to observe anything etcd related during scaling operations.
+In particular, it is the etcd operators responsibility to add Machine deletion hooks to prevent Control Plane Machines
+from being removed until they are no longer needed.
+When the `ControlPlaneMachineSet` controller observes that a newly created Machine is ready, it will delete the old
+Control Plane Machine signalling to the etcd operator to switch the membership of the etcd cluster between the old and
+new instances.
+Once the membership has been switched, the Machine deletion hook will be removed on the old Machine, allowing it to be
+removed by the Machine controller in the normal way.
+
+#### Removing/disabling the ControlPlaneMachineSet
+
+As some users may want to remove or disable the ControlPlaneMachineSet, a finalizer will be placed on the
+`ControlPlaneMachineSet` to allow the controller to ensure a safe removal of the ControlPlaneMachineSet, while leaving
+the Control Plane Machines in place.
+
+Notably it will need to ensure that there are no owner references on the Machines pointing to the ControlPlaneMachineSet
```

Comment:
It is not clear to me why we would even want an owner reference from the `ControlPlaneMachineSet` to the `Machines`. What is the use case for that?

Reply:
Having the owner reference signals to other controllers in the cluster that there is an owner for the Machine, in which case this changes the logic for those controllers.

Having an owner implies that if you change the object, it may be put back if your change conflicts with the desired spec (eg if a user goes and modifies the Machine, they will cause a rollout and for that Machine to be replaced, the owner will hint to them they shouldn't be modifying it I hope). 

It also implies that the object would be replaced should it be deleted, it will be recreated. This logic is used within the MachineHealthCheck, so adding an Owner Ref to the CPMS managed Machines then enables MHC to work with Control Plane Machines